File: notificare_flutter/ios/Classes/Events/NotificareEventManager.swift
//
//  NotificareEventManager.swift
//  notificare
//
//  Created by Helder Pinhal on 03/12/2020.
//

import Foundation

class NotificareEventManager {

    static let shared = NotificareEventManager()

    private var channels: [NotificareEventType: FlutterEventChannel] = [:]
    private var streams: [NotificareEventType: NotificareEventStream]

    private init() {
        var streams: [NotificareEventType: NotificareEventStream] = [:]

        NotificareEventType.allCases.forEach { type in
            streams[type] = NotificareEventStream(eventType: type)
        }

        self.streams = streams
    }

    func register(for registrar: FlutterPluginRegistrar) {
        streams.values.forEach { stream in
            if let channel = channels[stream.eventType] {
                channel.setStreamHandler(stream)
            } else {
                let channel = FlutterEventChannel(name: stream.name, binaryMessenger: registrar.messenger(), codec: FlutterJSONMethodCodec.sharedInstance())
                channel.setStreamHandler(stream)

                channels[stream.eventType] = channel
            }
        }
    }

    func unregister(for registrar: FlutterPluginRegistrar) {
        channels.values.forEach { channel in
            channel.setStreamHandler(nil)
        }
    }

    func send(_ event: NotificareEvent) {
        DispatchQueue.main.async { [weak self] in
            self?.streams[event.type]?.send(event)
        }
    }
}

File: notificare_flutter/android/src/main/kotlin/re/notifica/flutter/events/NotificareEventManager.kt
package re.notifica.flutter.events

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.JSONMethodCodec

internal object NotificareEventManager {

    private val channels = mutableMapOf<NotificareEvent.Type, EventChannel>()

    private val streams: Map<NotificareEvent.Type, NotificareEventStream> by lazy {
        NotificareEvent.Type.values().map {
            it to NotificareEventStream(it)
        }.toMap()
    }

    fun register(messenger: BinaryMessenger) {
        streams.forEach { (eventType, stream) ->
            val channel = channels[eventType]
                ?: EventChannel(messenger, stream.name, JSONMethodCodec.INSTANCE).also {
                    // Keep a copy of it.
                    channels[eventType] = it
                }

            channel.setStreamHandler(stream)
        }
    }

    fun unregister() {
        channels.forEach { it.value.setStreamHandler(null) }
    }

    fun send(event: NotificareEvent) {
        Handler(Looper.getMainLooper()).post {
            streams[event.type]?.emit(event)
        }
    }
}
